### DISCUSSION TITLE
Help with using mp_sched_scheduler

### DISCUSSION THREAD

**Discussion by Kyderio** [micropython/micropython#12929] [Core Development]
Hi, I am new to micropython and trying to port GPIO IRQ feature for ameba boards. This is my python script.
```
import machine

def button_interrupt(pin):
    if pin.value() == 0:  # Button is pressed (assuming the pull-up configuration)
        print("Button pressed")
    else:
        print("Button released")

button_pin = machine.Pin("PA_12", machine.Pin.IN, machine.Pin.PULL_UP)

button_pin.irq(handler=button_interrupt,trigger=machine.Pin.IRQ_RISING)
```

This is my objpin.c code where 

```
void gpio_demo_irq_handler(uint32_t id, gpio_irq_event event)
{
    // TODO: pass real user-defined handler
    printf("This is a demo irq handler");
}

STATIC mp_obj_t machine_pin_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args)
{
    enum
    {
        ARG_handler,
        ARG_trigger,
        ARG_hard
    };
    static const mp_arg_t allowed_args[] = {
        {MP_QSTR_handler, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE}},
        {MP_QSTR_trigger, MP_ARG_INT, {.u_int = IRQ_FALL | IRQ_RISE}},
        {MP_QSTR_hard, MP_ARG_BOOL, {.u_bool = false}},
    };

    // deinit the GPIO first
    pin_obj_t *self_in = MP_OBJ_TO_PTR(pos_args[0]);
    gpio_deinit((gpio_t *)&(self_in->obj));

    printf("This is a test for irq function\n");

    pin_irq_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);

    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);

    if (n_args > 1 || kw_args->used != 0)
    {
        // Configure IRQ.

        self->trigger = args[ARG_trigger].u_int;
        self->id = self_in->id;

        gpio_irq_init((gpio_irq_t *)&(self->obj), self->id, gpio_demo_irq_handler, 0); // set handler
        gpio_irq_set((gpio_irq_t *)&(self->obj), self->trigger, true);                 // Set Trigger mode
        gpio_irq_enable((gpio_irq_t *)&(self->obj));
    }

    return mp_const_none;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pin_irq_obj, 1, machine_pin_irq);
```
I have read up a bit, but im not sure how to use mp_sched_schedule to pass the user-defined handler into the c code 

```
void gpio_demo_irq_handler(uint32_t id, gpio_irq_event event)
{
    // TODO: pass real user-defined handler
    printf("This is a demo irq handler");
}
```

Can someone guide me towards the right direction? Thanks!

**Comment by jimmo** [comment]
You need to have some structure in RAM that can map the information you receive in the C IRQ handler (i.e. `id`) to the handler that was passed to the `pin.irq` function. On some ports this is done by having the handler be a member of the pin struct (but this means you have to make the pin struct non-const and therefore in bss not rom), on other ports there's just a table in memory.

Note that these handlers have to be "root pointers" (so the GC doesn't clean them up).

Look at e.g. stm32/extint.c for example (it uses `MP_REGISTER_ROOT_POINTER` to create a table of handlers).

But also worth looking at `shared/runtime/mpirq.c` which is supposed to be a common helper for implementing the `.irq` method correctly (this is supposed to return an irq object that can be used to retrieve the event flags etc). Not all ports use this (yet).

**Reply by jimmo** [reply]
Also `mpirq.c` handles the hard/soft and how to invoke the callback directly or via the scheduler.

**Reply by Kyderio** [reply]
Thanks so much for the fast response!  Let me give it a try :)
